### PR TITLE
track https://ring.exchange

### PR DESCRIPTION
**Pull request by dericwa**
**NOTE**

#### Please enable "Allow edits by maintainers" while putting up the PR.

---

> - If you would like to add a `volume` adapter please submit the PR [here](https://github.com/DefiLlama/adapters).
> - If you would like to add a `liquidations` adapter, please refer to [this readme document](https://github.com/DefiLlama/DefiLlama-Adapters/tree/main/liquidations) for details.

1. Once your adapter has been merged, it takes time to show on the UI. If more than 24 hours have passed, please let us know in Discord.
2. Sorry, We no longer accept fetch adapter for new projects, we prefer the tvl to computed from blockchain data, if you have trouble with creating a the adapter, please hop onto our discord, we are happy to assist you.
3. Please fill the form below  **only if the PR is for listing a new protocol** else it can be ignored/replaced with reason/details about the PR
4. **For updating listing info** It is a different repo, you can find your listing in this file: https://github.com/DefiLlama/defillama-server/blob/master/defi/src/protocols/data2.ts, you can  edit it there and put up a PR
5. Do not edit/push `package-lock.json` file as part of your changes, we use lockfileVersion 2, and most use v1 and using that messes up our CI
6. No need to go to our discord and announce that you've created a PR, we monitor all PRs and will review it asap

---
##### Name (to be shown on DefiLlama):
Ring Protocol

##### Twitter Link:
https://twitter.com/ProtocolRing

##### List of audit links if any:
https://github.com/RingProtocol/audits

##### Website Link:
https://ring.exchange/

##### Logo (High resolution, will be shown with rounded borders):
https://drive.google.com/drive/shared-with-me

##### Current TVL:
$67 millions USD

##### Treasury Addresses (if the protocol has treasury)
NA

##### Chain:
Blast

##### Short Description (to be shown on DefiLlama):
Ring is a new era in Decentralized Exchange revolutionizing asset utilization.

##### Category (full list at https://defillama.com/categories) *Please choose only one:
DEX

##### forkedFrom (Does your project originate from another project):
https://github.com/Uniswap/v2-core

##### methodology (what is being counted as tvl, how is tvl being calculated):
For each token, we use a token factory, called FewFactory to create a wrapped token. The asset is kept in wrapped token, while the wrapped token is used in DEFI protocols.

##### Github org/user (Optional, if your code is open source, we can track activity):
https://github.com/RingProtocol

### CODE REVIEW

**Comment by llamatester** [comment]
The adapter at projects/ring-protocol exports TVL: 
        
 
 
    blast                     67.57 M
    
    total                    67.57 M

File: projects/ring-protocol/index.js
Comment (g1nt0ki) [inline projects/ring-protocol/index.js:1]: I have added support for pricing wrapped tokens here: https://github.com/DefiLlama/defillama-server/commit/db50300f2a82da24c79ef0349393c756d8d06ef3#diff-f635a1d2d76aafd5b504caced903b4d7edb0968305040cd515271d8ba4181658R2

So, maybe this PR can be updated to count only the dex tvl? 

    const { uniTvlExport } = require('../helper/unknownTokens')

    module.exports = uniTvlExport('blast', '0x24F5Ac9A706De0cF795A8193F6AB3966B14ECfE6')
    
I am not sure if we should count wrapped tokens as tvl

Comment (dericwa) [reply projects/ring-protocol/index.js:1]: Good suggestion. Here are a few reasons why we calculate TVL in wrapped tokens:

1. The wrapped tokens factory is built by the Ring Protocol, as we aim to create an ecosystem maximizing asset utilization to attract dapps for integrating our wrapped tokens factory.

2. According to our design, the native yields generated by USDB and WETH in Blast are captured by our factory. We redistribute these yields across all dapps using this layer, emphasizing that it's not merely a set of wrapped token contracts without creating value.

3. Upon reviewing our architecture, the DEX is our initial go-to-market product built on top of the wrapped tokens factory. We also leverage this factory for our yield farming programs. And more 3rd party dapp integrations are coming soon. Therefore, using your code suggestions won't calculate the TVL derived from yield farming. If you compare TVL, the current DEX TVL is around 3m, whereas the wrapped tokens factory's TVL is around 75m.
I wish these reasons answered your question.

Comment (dericwa) [reply projects/ring-protocol/index.js:1]: Separately question: we do want to show Ring Swap with current TVL in Blast DEX category with correct TVL (~3m) data based on your suggestion.
Should we create a separate pull requests to do so?

Comment (dericwa) [reply projects/ring-protocol/index.js:1]: Also, thanks for adding the few factory in code base. With that merge, I believe the code for tracking few factory tvl should be simpler. Could you provide your suggestion to refactor this pull request?

Comment (dericwa) [reply projects/ring-protocol/index.js:1]: I just run your code snippet, and got the following result:
```
--- blast ---
fwUSDB                    1.75 M
fwWETH                    1.75 M
Total: 3.51 M 
```
I think the misunderstanding comes from that `few factory` is actually part of Ring Protocol, and we are building multiple products on top of it: dex, farming, lending etc. Here is the source code for `few factory`: https://github.com/RingProtocol/few-factory

**Comment by g1nt0ki** [comment]
hi, I am closing this in favor of https://github.com/DefiLlama/DefiLlama-Adapters/commit/d003e75f7a4f5532a8a0219dba477a1a2bc5acca which does something similar

**Comment by dericwa** [comment]
> hi, I am closing this in favor of [d003e75](https://github.com/DefiLlama/DefiLlama-Adapters/commit/d003e75f7a4f5532a8a0219dba477a1a2bc5acca) which does something similar

WOW! I really appreciate this! Your code is way simpler than mine, and you did both for `swap` and `few`. Thank you!
Glad to work with you on this. Have a great day. 👍

**Comment by g1nt0ki** [comment]
> > hi, I am closing this in favor of [d003e75](https://github.com/DefiLlama/DefiLlama-Adapters/commit/d003e75f7a4f5532a8a0219dba477a1a2bc5acca) which does something similar
> 
> WOW! I really appreciate this! Your code is way simpler than mine, and you did both for `swap` and `few`. Thank you! Glad to work with you on this. Have a great day. 👍

Yeah, it is  easier for me since I am bit more used to this repo ;)
sometimes I feel bad for closing PRs as people put effort into it, sorry about that. Appreciate the PR

**Comment by dericwa** [comment]
> > > hi, I am closing this in favor of [d003e75](https://github.com/DefiLlama/DefiLlama-Adapters/commit/d003e75f7a4f5532a8a0219dba477a1a2bc5acca) which does something similar
> > 
> > 
> > WOW! I really appreciate this! Your code is way simpler than mine, and you did both for `swap` and `few`. Thank you! Glad to work with you on this. Have a great day. 👍
> 
> Yeah, it is easier for me since I am bit more used to this repo ;) sometimes I feel bad for closing PRs as people put effort into it, sorry about that. Appreciate the PR

The code is actually shocking. You found our two factories without asking. And you include `RING` and `RNG` which we actually haven't launched yet (but the symbols are correct). This saves lots of our future efforts too. That is why I am very surprised. Ty!

**Pull request closed without merging**